api/serialization/parquet.go

```diff
@@ -102,13 +103,35 @@ func (d *Parquet) ReadData(uri string) ([][]string, error) {
 	colCount := pr.SchemaHandler.GetColumnNum()
 	rowCount := pr.GetNumRows()
 	dataByCol := make([][]string, colCount)
+	var rLimit syscall.Rlimit
+	syscErr := syscall.Getrlimit(syscall.RLIMIT_NOFILE, &rLimit)
+	if syscErr != nil {
+		return nil, errors.Wrap(syscErr, "unable to get ulimit -n")
+	}
+	// rLimit.Cur = 25000
 	for i := int64(0); i < colCount; i++ {
+		if i > 0 && i % (int64(rLimit.Cur) - 100) == 0 {
+			fr.Close()
+			pr.ReadStop()
+			fr, err = local.NewLocalFileReader(uri)
+			if err != nil {
+				fr.Close()
+				return nil, errors.Wrap(err, "unable to open parquet file")
+			}
+			pr, err = reader.NewParquetColumnReader(fr, 1)
+			if err != nil {
+				pr.ReadStop()
+				return nil, errors.Wrap(err, "unable to open parquet file reader")
+			}
+		}
 		colRaw, err := d.readColumn(pr, i, rowCount)
 		if err != nil {
 			return nil, err
 		}
 		dataByCol[i] = d.columnToString(colRaw, *pr.SchemaHandler.SchemaElements[i+1].Type)
 	}
+	defer fr.Close()
```

Comment:
I think these defers are misplaced. At this point, they can probably just be closed. However, if an error occurs reading the column, then they are never closed. I think these defers should be right after opening them. Although since the previously opened one gets closed explicitly, then the whole thing gets a bit messy.

Reply:
I will just take out the defers for now, then